components/builder-events/src/kafka.rs

```diff
@@ -27,43 +37,112 @@ pub struct KafkaProducer {
+                      message_timeout_ms: 6000,
+                      partition_eof:      false,
+                      session_timeout_ms: 6000, }
+    }
+}
+
+impl TryFrom<&KafkaConfig> for KafkaConsumer {
+    type Error = Error;
+
+    fn try_from(config: &KafkaConfig) -> Result<Self, Error> {
+        let bootstrap_list = config.bootstrap_nodes.join(",");
+        match ClientConfig::new().set("group.id", &config.group_id)
+                                 .set("bootstrap.servers", &bootstrap_list)
+                                 .set("enable.partition.eof", &config.partition_eof.to_string())
+                                 .set("sasl.username", &config.api_key)
+                                 .set("sasl.password", &config.api_secret_key)
+                                 .set("sasl.mechanisms", "PLAIN")
+                                 .set("security.protocol", "SASL_SSL")
+                                 .set("session.timeout.ms", &config.session_timeout_ms.to_string())
+                                 .set("enable.auto.commit", "true")
+                                 .set("auto.offset.reset", "earliest")
+                                 .create()
+        {
+            Ok(c) => Ok(KafkaConsumer { inner: c }),
+            Err(err) => {
+                error!("Error initializing kafka consumer: {:?}", err);
+                Err(Error::EventBusError(Box::new(err)))
+            }
+        }
     }
 }
 
 impl TryFrom<&KafkaConfig> for KafkaProducer {
-    type Error = KafkaError;
+    type Error = Error;
 
-    fn try_from(config: &KafkaConfig) -> Result<Self, KafkaError> {
+    fn try_from(config: &KafkaConfig) -> Result<Self, Error> {
         let bootstrap_list = config.bootstrap_nodes.join(",");
         match ClientConfig::new().set("bootstrap.servers", &bootstrap_list)
                                  .set("message.timeout.ms", &config.message_timeout_ms.to_string())
                                  .set("client.id", &config.client_id.as_str())
                                  .set("sasl.username", &config.api_key)
                                  .set("sasl.password", &config.api_secret_key)
-                                 .set("security.protocol", "SASL_SSL")
                                  .set("sasl.mechanisms", "PLAIN")
+                                 .set("security.protocol", "SASL_SSL")
                                  .create()
         {
             Ok(p) => Ok(KafkaProducer { inner: p }),
             Err(err) => {
                 error!("Error initializing kafka producer: {:?}", err);
-                Err(err)
+                Err(Error::EventBusError(Box::new(err)))
+            }
+        }
+    }
+}
+
+fn extract_message_payload(msg: &OwnedMessage) -> String {
+    match msg.payload_view::<str>() {
+        Some(Ok(payload)) => payload.to_string(),
+        Some(Err(_)) => "Message payload is not a string".to_owned(),
+        None => "No payload".to_owned(),
```

Comment:
In hindsight, this is unintentional. Adjusted in https://github.com/habitat-sh/builder/pull/1549/commits/261fcc5a3138e104852fa8476e681b947820d042

Reply:
My initial (poorly-expressed; sorry!) point was a little broader than this. The output of this function is a `String`, but that string could be any of the following:

- an actual message
- "Message payload is not a string"
- "No payload"

There is no way that the caller of this method can distinguish between any of these cases based on what they receive. To be very concrete (and I doubt it would ever come up in practice, but still), what are we to think about a message with a payload of "No payload"? :thinking: 

It seems like you're trying to get something general and Builder-specific out of an `rdkafka` crate type, where that "something general" is currently a `String` (though you might imagine something more domain specific in the future, but I digress :sweat_smile: ). It seems that it may still be useful to carry the payload / error / no-payload distinction forward. 

I'm not sure what that might eventually be, but you could minimally capture it by returning `Option<Result<String>>` here (or perhaps just `Result<String>`, since (for Builder's purposes) it may actually be an error to have no payload).

Reply:
I agree with your assessment and opted for a `Option<Result<String>>` signature. That way we can capture the fact that there was a message via the `Option` and the integrity of the pulled message via the `Result`.  This is in https://github.com/habitat-sh/builder/pull/1549/commits/8e47c6a7332931a9ed7cd6b2976fce410d83677c